File: C++/프로그래머스/0/120866. 안전지대/안전지대.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

void check_bomb_area(vector<vector<int>>& b, int x, int y, int& count) {
    int n = b.size();
    
    // 폭탄(현재 칸) count에서 제거
    count--;
    
    // x, y 기준 반경 1칸(총 8칸) 순회
    for (int r = y - 1; r < y + 2; r++) {
        for (int c = x - 1; c < x + 2; c++) {
            // board의 범위 내이고 해당 칸이 0이면
            if (r > -1 && r < n && c > -1 && c < n && b[r][c] == 0) {
                // 체크하고 count 감소
                b[r][c] = -1;
                count--;
            }
        }
    }
    //순회 끝
}

int solution(vector<vector<int>> board) {
    int n = board.size();
    vector<vector<int>> c_board(board);
    
    // 전체 칸의 갯수
    int answer = n * n;
    
    for (int y = 0; y < n; y++) {
        for (int x = 0; x < n; x++) {
            // 이미 체크된(-1) 칸은 pass
            if (c_board[y][x] == -1) continue;
            // 폭탄이 있으면 check_bomb_area 실행
            if (c_board[y][x] == 1)
                check_bomb_area(c_board, x, y, answer);
        }
    }
    
    return answer;
}

File: Go/프로그래머스/0/120866. 안전지대/안전지대.go
// (x, y) 주변을 위험 지역으로 표기하고 표기한 칸수 반환
func CheckDanger(b [][]int, x, y int) int {
    // 현재 칸 포함
    count := 1
    
    // x, y의 최댓값
    n := len(b)
    for r := y-1; r <= y+1; r++ {
        for c := x-1; c <= x+1; c++ {
            // 좌표가 최대/최소 내에 있고, 0으로 표기되어 있으면
            if r > -1 && r < n && c > -1 && c < n && b[r][c] == 0 {
                // 위험 지역 처리 후 count++
                b[r][c] = -1
                count++
            }
        }
    }
    
    return count
}

func solution(board [][]int) int {
    res := len(board) * len(board[0])
    
    // 원본 슬라이스 훼손 방지를 위한 복사
    c_board := make([][]int, len(board))
    
    for i, row := range board {
        c_row := make([]int, len(row))
        copy(c_row, row)
        c_board[i] = c_row
    }
    
    for y, row := range c_board {
        for x, v := range row {
            if v == 1 {
                res -= CheckDanger(c_board, x, y)
            }
        }
    }
    
    return res
}